File: src/competitve practice/leetcode-30day-challenge/week4/LRUCache.java

class LRUCache {
    class Node {
        int key;
        int value;
        Node next;
        Node prev;
    }

    Node head = new Node();
    Node tail = new Node();
    Map<Integer, Node> map;
    int cap;

    public LRUCache(int capacity) {
        this.map = new HashMap<Integer, Node>(capacity);
        this.cap = capacity;
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        int result = -1;
        Node node = map.get(key);
        if (node != null) {
            removeNode(node);
            addNode(node);
            result = node.value;
        }
        return result;
    }

    public void put(int key, int value) {
        Node node = map.get(key);

        if (node != null) {
            removeNode(node);
            node.value = value;
            addNode(node);
        } else {
            if (map.size() == cap) {
                Node toRemove = tail.prev;
                map.remove(toRemove.key);
                removeNode(toRemove);
            }

            node = new Node();
            node.key = key;
            node.value = value;
            map.put(key, node);
            addNode(node);
        }
    }

    public void addNode(Node node) {
        node.next = head.next;
        head.next.prev = node;
        head.next = node;
        node.prev = head;
    }

    public void removeNode(Node node) {
        Node next = node.next;
        Node prev = node.prev;
        next.prev = prev;
        prev.next = next;
    }

}

/**
 * Your LRUCache object will be instantiated and called as such: LRUCache obj =
 * new LRUCache(capacity); int param_1 = obj.get(key); obj.put(key,value);
 */

class LRUCache2 {
    Map<Integer, Integer> map;
    List<Integer> list;
    int capacity;

    public LRUCache2(int capacity) {
        this.map = new HashMap<Integer, Integer>();
        this.list = new ArrayList<Integer>();
        this.capacity = capacity;
    }

    public int get(int key) {
        int result = -1;
        if (map.containsKey(key) && list.contains(key)) {
            result = map.get(key);
            list.remove(list.indexOf(key));
            list.add(key);
        }
        return result;
    }

    public void put(int key, int value) {
        if (list.contains(key))
            list.remove(list.indexOf(key));

        if (list.size() < capacity) {
            list.add(key);
        } else {
            list.remove(0);
            list.add(key);
        }
        map.put(key, value);
    }
}

File: LeetCode/leetcode-30day-challenge/week4/C++/LRUCache.cpp
#include <bits/stdc++.h>
using namespace std;

//Type: Implementation, Linked List, Maps
//Approach: We shall maintain a linked list of nodes,
//          and keep adding elements at the head of list
//          Now, to have access of the elements in O(1) time,
//          we will have to use a map data structure,
//          which will store the ptr to the node corresponding to the key.
//          The back of the linkedlist will contain the least recently used nodes.
//Complexity: O(1) time

class LRUCache
{
public:
    struct Node
    {
        int key;
        int value;
        Node *next;
        Node *prev;
        Node() : next(nullptr), prev(nullptr) {}
        Node(int k, int v) : key(k), value(v), next(nullptr), prev(nullptr) {}
    };

    Node *head = new Node;
    Node *tail = new Node;
    unordered_map<int, Node *> nodeMap;
    int cap;

    void addNode(Node *node)
    {
        node->next = head->next;
        head->next->prev = node;
        head->next = node;
        node->prev = head;
    }

    void removeNode(Node *node)
    {
        Node *next = node->next;
        Node *prev = node->prev;
        next->prev = prev;
        prev->next = next;
    }

    LRUCache(int capacity)
    {
        head->next = tail;
        tail->prev = head;
        cap = capacity;
    }

    int get(int key)
    {
        int result = -1;
        if (nodeMap.count(key))
        {
            removeNode(nodeMap[key]);
            addNode(nodeMap[key]);
            result = nodeMap[key]->value;
        }
        return result;
    }

    void put(int key, int value)
    {
        if (nodeMap.count(key))
        {
            removeNode(nodeMap[key]);
            nodeMap[key]->value = value;
            addNode(nodeMap[key]);
        }
        else
        {
            if (nodeMap.size() == cap)
            {
                Node *toRemove = tail->prev;
                nodeMap.erase(toRemove->key);
                removeNode(toRemove);
            }
            Node *newNode = new Node{key, value};
            nodeMap[key] = newNode;
            addNode(newNode);
        }
    }
};